docs/design/generic.md

```diff
@@ -0,0 +1,95 @@
+# Generic artifact fetching
+
+## Introduction
+
+This document will describe high-level implementation overview for supporting generic artifact fetching in cachi2.
+Up until now cachi2 has only supported package managers for various ecosystems and languages.
+However, there are a couple of use-cases where language non-specific artifacts need to be pre-fetched in order to satisfy
+requirements of a hermetic build.
+
+## Context
+
+For context, generic artifact fetching is a use-case of its own (e.g. [OVAL feeds](https://github.com/CISecurity/OVALRepo),
+AI models), it is also necessary precursor for implementing support for fetching maven artifacts, which won't be covered
+in this design, but in a followup document.
+
+## Design
+
+In this section, I will try to cover individual parts of the design.
+
+### Source repository
+
+This section will describe the structure of the source repository, that will serve as an input to cachi2. The idea is to
+define a cachi2 lockfile that will specify individual artifacts to fetch along with necessary metadata - e.g. checksums.
+The format chosen for this lockfile is yaml, and will include [purl](https://github.com/package-url/purl-spec) for each
+of the fetched artifacts. This decision was made mainly because it allows for followup implementation of maven support,
+with accurate SBOM information. Here's an example of such a lockfile.
+
+```yaml
+artifacts:
+  - purl: pkg:generic/granite-model?download_url=https://huggingface.co/instructlab/granite-7b-lab/resolve/main/model-00001-of-00003.safetensors?download=true
+    target: granite-model.safetensors
+    checksums:
+      sha256: 07123e1f482356c415f684407a3b8723e10b2cbbc0b8fcd6282c49d37c9c1abc
+```
+
+#### Lockfile format and validation
+
+##### purl (required)
+
+At this point, the only purl type allowed would be `pkg:generic`. This is because cachi2 has no good way of verifying
+additional properties of the fetched artifact that could be included in the resulting SBOM. This should create a strong
+incentive to use this feature in the only truly necessary cases, because it will generate low-quality SBOM components,
+as compared to using other package managers provided by cachi2. Additionally, the only allowed qualifier should be `download_url`.
+
+#### target (optional)
+
+This is mainly for the users convenience, so the files end up in expected locations. Target here means a specific subdirectory
+inside cachi2's output directory. Special care needs to be taken to ensure there is not a conflict with other downloaded files.
+If not specified, filename of the downloaded file will be used.
+
+##### checksums (optional)
+
+I've chosen tho separate checksums from the purl, mostly for better readability of the lockfile, but this can be up for
+discussion. If no checksum is provided, cachi2 should still download the artifact, but report this fact in the output
```

Comment:
I'm not sure if this is the best approach here. We have precedent in allowing artifacts to be downloaded without user-defined checksums, but we'd still have the server-provided checksums. If no checksum is verified at all, I believe it brings a greater risk for the build that will consume the prefetched content.

Reply:
> The idea here is delegate the decision to Enterprise Contract policy, to manage that risk, but be transparent about unverified files.

Please don't mention specific components where this is consumed, at this point in time this project isn't strictly tied to the Konflux project and so if you need to make an argument in a community space then try to make it true in general, e.g. `It cannot be assumed in general that each type or artifact will always have a corresponding checksum server-side and hence the missing checksum needs to be reported in the SBOM and verified in a different software supply chain pipeline stage`.

> > This is a requirement in the feature doc ([non-public link](https://docs.google.com/document/d/15Ibevfwp7FoafaBQcbtxYYJN-aoiF5jq8aNloA8UIZg/edit#bookmark=id.n7vnmxsap2sj))

Please don't paste links to internal documents in public spaces! You wouldn't do this in Bugzilla I presume without first marking the Bugzilla private. Annotating this as a private resource is as helpful to the context as not having the link at all, so either share a public version of it or proceed over a private channel, but not a public community space.

Reply:
Coming back to the original point, after a chat with other maintainers, we feel that the checksum validation is a very important step in the prefetch. At least one checksum should be validated for each downloaded artifact, and since for the generic fetcher we won't have server-provided checksums, we should just require the user to provide them.

This is extra sensitive for the generic fetcher: we are not relying on an already stablished language ecosystem, which means the user could download anything from the internet. We can't avoid that malicious content is fetched, but we should at least have strong guarantees that the content downloaded was exactly what the user expected.

Reply:
We'll need to hash it out (pun intended) with the Konflux people, but generally, I don't see this as a blocker to moving ahead with the overall implementation. 

I'm leaning towards agreeing with you, however, I'll need to confirm this is not a hard requirement.

Reply:
Alright, please share your findings.

Reply:
I will. For now I've changed the design to require checksums.